modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/disk/AbstractExternalResult.java

```diff
@@ -0,0 +1,374 @@
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.query.h2.disk;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.concurrent.atomic.AtomicLong;
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.IgniteException;
+import org.apache.ignite.IgniteLogger;
+import org.apache.ignite.internal.GridKernalContext;
+import org.apache.ignite.internal.processors.cache.persistence.file.FileIO;
+import org.apache.ignite.internal.processors.cache.persistence.file.FileIOFactory;
+import org.apache.ignite.internal.processors.query.h2.H2MemoryTracker;
+import org.apache.ignite.internal.processors.query.h2.H2Utils;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.h2.result.ResultExternal;
+import org.h2.result.ResultInterface;
+import org.h2.store.Data;
+import org.h2.value.Value;
+import org.jetbrains.annotations.NotNull;
+
+import static java.nio.file.StandardOpenOption.CREATE_NEW;
+import static java.nio.file.StandardOpenOption.READ;
+import static java.nio.file.StandardOpenOption.WRITE;
+import static org.apache.ignite.internal.processors.query.h2.IgniteH2Indexing.DISK_SPILL_DIR;
+
+/**
+ * Basic class for external result. Contains common methods for file IO.
+ */
+@SuppressWarnings({"MissortedModifiers", "WeakerAccess", "ForLoopReplaceableByForEach"})
+public abstract class AbstractExternalResult implements ResultExternal {
+    /** File name generator. */
+    private static final AtomicLong idGen = new AtomicLong();
+
+    /** Initial rows byte buffer size. */
+    protected static final int INIT_BUF_SIZE = 512;
+
+    /** Serialized row header size. */
+    protected static final int ROW_HEADER_SIZE = 8; // 2 x int: row length in bytes + column count.
+
+    /** Logger. */
+    protected final IgniteLogger log;
+
+    /** Current size in rows. */
+    protected int size;
+
+    /** File. */
+    protected final File file;
+
+    /** Spill file IO.*/
+    protected volatile FileIO fileIo;
+
+    /** Memory tracker. */
+    protected final H2MemoryTracker memTracker;
+
+    /** Parent result. */
+    protected final AbstractExternalResult parent;
+
+    /** Child results count. Parent result is closed only when all children are closed. */
+    private int childCnt;
+
+    /** */
+    private boolean closed;
+
+    /**
+     * @param ctx Kernal context.
+     * @param memTracker Memory tracker
+     */
+    protected AbstractExternalResult(GridKernalContext ctx, H2MemoryTracker memTracker) {
+        this.log = ctx.log(AbstractExternalResult.class);
+
+        try {
+            String fileName = "spill_" + ctx.localNodeId() + "_" + idGen.incrementAndGet();
+
+            file = new File(U.resolveWorkDirectory(
+                ctx.config().getWorkDirectory(),
+                DISK_SPILL_DIR,
+                false
+            ), fileName);
+
+            file.deleteOnExit();
```

Comment:
This causes put into shared LinkedHashSet inside synchronized block.
Let's remove it as 
1. directory will be cleaned on next node start
2. it is designed that fileName will never be reused (filename contains node uuid).

Reply:
Done